### ISSUE TITLE
Allow EagerLoader to accept any callable as query builder

### ISSUE THREAD

**Issue by AidanHaddonWright** [cakephp/cakephp#13337]
This is a (multiple allowed):

* [ ] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.7.7

I would like to suggest, what I hope is, a simple edit to the EagerLoader class.
I had a look to see if this has been raised before but I couldn't find anything so sorry if this is a duplicate.

On [this](https://github.com/cakephp/cakephp/blob/master/src/ORM/EagerLoader.php#L435) line, is it possible to change it to:
```php
if (is_callable($options)) {
    $options = ['queryBuilder' => $options];
}
```

Basically, in an effort to make my code more readable, I wanted to move what was a closure into its own callable class so it can be reused. E.g:

````php
// Before
$query->contain(['SomeAssociation' => function (Query $query) {
        $query->foo();
        $query->bar();
        return $query;
}]);

//After
class FooBarQueryCallback {
    public function __invoke(Query $query)
    {
        $query->foo();
        $query->bar();
        return $query;
    }
}

$query->contain(['SomeAssociation' => new FooBarQueryCallback]);
```

Ignoring the methods used, this code won't work because EagerLoader only accepts Closures (which cannot be extended).

I know I could use a behavior with a finder in it but I feel like that's overkill for a function that will only be used a limited number of times.

**Comment by ADmad** [comment]
If you are using PHP 7.1+ you can use

```php
['SomeAssociation' => \Closure::fromCallable(new FooBarQueryCallback)]
```

**Comment by ADmad** [comment]
Coincidentally I was reading the RFC for "Closure from callable function" and using closures instead of callables is more [beneficial](https://wiki.php.net/rfc/closurefromcallable#why_would_you_use_this). So I would rather not change this.

**Comment by AidanHaddonWright** [comment]
Makes sense.

I didn't realise it had to validate a callable every time it was called...

**Comment by markstory** [comment]
Going to close this then as the requested pattern can be achieved using Closure::fromcCallable().

**Comment by AidanHaddonWright** [comment]
In that case, I noticed that the CollectionTrait mostly uses callable instead of Closure. Infact it was the [collections documentation](https://book.cakephp.org/3.0/en/core-libraries/collections.html#making-collection-methods-reusable) where I learned about the __invoke method.

Depending on your plans for supporting legacy PHP versions in 4.x, would it be worth updating Collections to also use Closure (as mentioned the method above only works as of PHP 7.1).

In theory it would increase the speed of iterating over large collections.
